Inductor cpp wrapper: cache the loading of the kernel

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89743
* #90755
* __->__ #89742

### Pitch
Cache the loaded kernel to reduce the overhead.

#### Code before:
```cpp
std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
    assert(kernel_cpp_0_lib != nullptr);
    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
    *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

#### Code after:
```cpp
template <typename KernelFunc>
KernelFunc load_cpp_kernel(const char* so_filename) {
    KernelFunc kernel_cpp;
    auto kernel_cpp_lib = dlopen(so_filename, RTLD_NOW);
    assert(kernel_cpp_lib != nullptr);
    *(void **) (&kernel_cpp) = dlsym(kernel_cpp_lib, "kernel");
    return kernel_cpp;
}


std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    static auto kernel_cpp_0 = load_cpp_kernel<void (*)(const float*,const float*,float*,float*)>("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so");
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

chunyuan-w committed:
Inductor cpp wrapper: cache the loading of the kernel

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 reviewed (commented):
Is it simpler to create a common cpp kernel loader class that can work with given persistent kernel library? Creating a new class for each kernel instance looks awkward.
```C++
template <typename KernelFunc>
KernelFunc load_cpp_kernel(const char* so_filename);

... call(...) {
  static auto kernel = cpp_kernel_loader<void (*)(const float*,const float*,float*,float*)>("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so");
  kernel(...);
   ...
}
```

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"


### Pitch
Cache the loaded kernel to reduce the overhead.

#### Code before:
```cpp
std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
    assert(kernel_cpp_0_lib != nullptr);
    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
    *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

#### Code after:
```cpp
class LoadKernel_call0 {
  public:
    LoadKernel_call0() {
        auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
        assert(kernel_cpp_0_lib != nullptr);
        *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    }

    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
};


std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    static LoadKernel_call0 load_kernel_;
    load_kernel_.kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"


### Pitch
Cache the loaded kernel to reduce the overhead.

#### Code before:
```cpp
std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
    assert(kernel_cpp_0_lib != nullptr);
    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
    *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

#### Code after:
```cpp
class LoadKernel_call0 {
  public:
    LoadKernel_call0() {
        auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
        assert(kernel_cpp_0_lib != nullptr);
        *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    }

    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
};


std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    static LoadKernel_call0 load_kernel_;
    load_kernel_.kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"


### Pitch
Cache the loaded kernel to reduce the overhead.

#### Code before:
```cpp
std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
    assert(kernel_cpp_0_lib != nullptr);
    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
    *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

#### Code after:
```cpp
class LoadKernel_call0 {
  public:
    LoadKernel_call0() {
        auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
        assert(kernel_cpp_0_lib != nullptr);
        *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    }

    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
};


std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    static LoadKernel_call0 load_kernel_;
    load_kernel_.kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"


### Pitch
Cache the loaded kernel to reduce the overhead.

#### Code before:
```cpp
std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
    assert(kernel_cpp_0_lib != nullptr);
    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
    *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

#### Code after:
```cpp
class LoadKernel_call0 {
  public:
    LoadKernel_call0() {
        auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
        assert(kernel_cpp_0_lib != nullptr);
        *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    }

    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
};


std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    static LoadKernel_call0 load_kernel_;
    load_kernel_.kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w commented:
> Is it simpler to create a common cpp kernel loader class that can work with given persistent kernel library? Creating a new class for each kernel instance looks awkward.
> 
> ```c++
> template <typename KernelFunc>
> KernelFunc load_cpp_kernel(const char* so_filename);
> 
> ... call(...) {
>   static auto kernel = cpp_kernel_loader<void (*)(const float*,const float*,float*,float*)>("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so");
>   kernel(...);
>    ...
> }
> ```

Refactored the code to follow the above suggestion.

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"


### Pitch
Cache the loaded kernel to reduce the overhead.

#### Code before:
```cpp
std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
    assert(kernel_cpp_0_lib != nullptr);
    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
    *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

#### Code after:
```cpp
class LoadKernel_call0 {
  public:
    LoadKernel_call0() {
        auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
        assert(kernel_cpp_0_lib != nullptr);
        *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    }

    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
};


std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    static LoadKernel_call0 load_kernel_;
    load_kernel_.kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: cache the loading of the kernel"


### Pitch
Cache the loaded kernel to reduce the overhead.

#### Code before:
```cpp
std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    auto kernel_cpp_0_lib = dlopen("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so", RTLD_NOW);
    assert(kernel_cpp_0_lib != nullptr);
    void (*kernel_cpp_0)(const float*,const float*,float*,float*);
    *(void **) (&kernel_cpp_0) = dlsym(kernel_cpp_0_lib, "kernel");
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

#### Code after:
```cpp
template <typename KernelFunc>
KernelFunc load_cpp_kernel(const char* so_filename) {
    KernelFunc kernel_cpp;
    auto kernel_cpp_lib = dlopen(so_filename, RTLD_NOW);
    assert(kernel_cpp_lib != nullptr);
    *(void **) (&kernel_cpp) = dlsym(kernel_cpp_lib, "kernel");
    return kernel_cpp;
}


std::vector<at::Tensor> call_0(std::tuple<at::Tensor&, at::Tensor&> args) {
    ...
    static auto kernel_cpp_0 = load_cpp_kernel<void (*)(const float*,const float*,float*,float*)>("/tmp/torchinductor_xxx/yr/cyr3uymlc6pgvnimx3fnynaa4t7ldafeqzhe5zpizmvorisx4hb2.so");
    kernel_cpp_0((float*)(arg0_1.data_ptr()), (float*)(arg1_1.data_ptr()), (float*)(buf0.data_ptr()), (float*)(buf1.data_ptr()));
    ...
}
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge
